app/src/main/java/br/com/catbag/gifreduxsample/asyncs/downloader/FileDownloader.java

```diff
@@ -0,0 +1,125 @@
+package br.com.catbag.gifreduxsample.asyncs.downloader;
+
+import android.util.Log;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import okhttp3.OkHttpClient;
+import okhttp3.ResponseBody;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+import retrofit2.http.GET;
+import retrofit2.http.Streaming;
+import retrofit2.http.Url;
+
+/**
+ * Created by felipe on 13/10/16.
+ */
+
+public class FileDownloader {
+
+    private static DownloadRoute sRoutes;
+
+    public FileDownloader() {
+        if (sRoutes == null) {
+            sRoutes = new RoutesGenerator().createRoutes(DownloadRoute.class);
+        }
+    }
+
+    public void download(String fileUrl, String pathToSave,
+                         SuccessDownloadListener successDownloadListener,
+                         FailureDownloadListener failureDownloadListener) {
+        Call<ResponseBody> call = sRoutes.downloadFileWithDynamicUrlSync(fileUrl);
+        call.enqueue(new Callback<ResponseBody>() {
+            @Override
+            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
+                if (response.isSuccessful()) {
+                    new Thread(() -> {
```

Comment:
If we uses the retrofit callback executors we can always ensure that responses will not running in main thread. 
However, when any executors are created the response will run in mainThread.

Ref: https://square.github.io/retrofit/2.x/retrofit/retrofit2/Callback.html